modules/siddhi-core/src/main/java/io/siddhi/core/table/CacheTableFIFO.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.siddhi.core.table;
+
+import io.siddhi.core.config.SiddhiAppContext;
+import io.siddhi.core.event.ComplexEvent;
+import io.siddhi.core.event.stream.StreamEvent;
+import io.siddhi.core.table.holder.IndexEventHolder;
+import io.siddhi.query.api.definition.Attribute;
+import io.siddhi.query.api.definition.TableDefinition;
+import org.apache.log4j.Logger;
+
+import java.util.Arrays;
+import java.util.Set;
+
+import static io.siddhi.core.util.SiddhiConstants.CACHE_TABLE_TIMESTAMP_ADDED;
+
+/**
+ * cache table with FIFO entry removal
+ */
+public class CacheTableFIFO extends CacheTable {
+    private static final Logger log = Logger.getLogger(CacheTableFIFO.class);
+
+    @Override
+    void addRequiredFieldsToCacheTableDefinition(TableDefinition cacheTableDefinition, boolean cacheExpiryEnabled) {
+        cacheTableDefinition.attribute(CACHE_TABLE_TIMESTAMP_ADDED, Attribute.Type.LONG);
+        cachePolicyAttributePosition = cacheTableDefinition.getAttributeList().size() - 1;
+        numColumns = cachePolicyAttributePosition + 1;
+    }
+
+    @Override
+    public void deleteOneEntryUsingCachePolicy() {
+        try {
+            IndexEventHolder indexEventHolder = (IndexEventHolder) stateHolder.getState().getEventHolder();
+            Set<Object> keys = indexEventHolder.getAllPrimaryKeyValues();
+            long minTimestamp = Long.MAX_VALUE;
+            Object keyOfMinTimestamp = null;
+            for (Object key: keys) {
+                Object[] data = indexEventHolder.getEvent(key).getOutputData();
+                long timestamp = (long) data[cachePolicyAttributePosition];
+                if (timestamp < minTimestamp) {
+                    minTimestamp = timestamp;
+                    keyOfMinTimestamp = key;
+                }
+            }
+            indexEventHolder.deleteEvent(keyOfMinTimestamp);
+        } catch (ClassCastException e) {
+            log.error(siddhiAppContext + ": " + e.getMessage());
+        }
+    }
+
+    @Override
+    public void deleteEntriesUsingCachePolicy(int numRowsToDelete) {
+        try {
+            IndexEventHolder indexEventHolder = (IndexEventHolder) stateHolder.getState().getEventHolder();
+            Set<Object> keys = indexEventHolder.getAllPrimaryKeyValues();
+            long[] minTimestampArray = new long[numRowsToDelete];
+            Arrays.fill(minTimestampArray, Long.MAX_VALUE);
+            Object[] keyOfMinTimestampArray = new Object[numRowsToDelete];
+
+            for (Object key: keys) {
+                Object[] data = indexEventHolder.getEvent(key).getOutputData();
```

Comment:
I'm not sure whether this logic is correct. 
You have to use a Tree structure here first add `numRowsToDelete` amount of events and then for every add remove the newest time.
Then you will get the list of `numRowsToDelete` amount of oldest events